libs/core/langchain_core/tools/convert.py

```diff
@@ -73,8 +81,8 @@ def tool(
     parse_docstring: bool = False,
     error_on_invalid_docstring: bool = True,
 ) -> Union[
-    BaseTool,
-    Callable[[Union[Callable, Runnable]], BaseTool],
+    Union[BaseTool, property],
```

Comment:
I think this type annotation shouldn't need to change. if we output a BaseTool, won't that get set as the name?

Reply:
Unfortunately mypy has different opinions.
Jokes aside, what about this change has you concerned? The return type has changed because method tools have type `property`. They take in a class instance (`self` value) and return a `StructuredTool`.

Reply:
I believe adding `property` to this unfortunately loosens the contract for @tool, such that static type checkers like mypy will not be able to determine that the attribute is now a `BaseTool`, which could be an annoying change for folks who are using this on regular functions (not methods)

Reply:
Its true, I can’t think of any other way that this can be implemented with:
1. Using the tool decorator
2. Not having to use function call syntax

I’m going to resolve this thread also and continue the conversation in the other thread, since the implementation is likely going to have to change significantly if these are unacceptable issues.